Libplanet/Assets/FungibleAssetValue.cs

```diff
@@ -0,0 +1,530 @@
+{
+    /// <summary>
+    /// Holds a fungible asset value which holds its <see cref="Currency"/> together.
+    /// <para>It behaves like numbers except for division operator (<c>/</c>) to prevent to forget
+    /// to handle its remainder; use <see cref="DivRem(FungibleAssetValue)"/> and <see
+    /// cref="DivRem(BigInteger)"/> methods instead.</para>
+    /// </summary>
+    [Serializable]
+    public readonly struct FungibleAssetValue :
+        IEquatable<FungibleAssetValue>,
+        IComparable<FungibleAssetValue>,
+        IComparable,
+        ISerializable
+    {
+        /// <summary>
+        /// The currency of the fungible asset.
+        /// </summary>
+        public readonly Currency Currency;
+
+        internal readonly BigInteger RawValue;
+
+        /// <summary>
+        /// Creates a zero value of the <paramref name="currency"/>.
+        /// </summary>
+        /// <param name="currency">The currency to create a zero value.</param>
+        public FungibleAssetValue(Currency currency)
+            : this(currency, BigInteger.Zero)
+        {
+        }
+
+        /// <summary>
+        /// Creates a value of the <paramref name="currency"/> from the given
+        /// <paramref name="majorUnit"/> and <paramref name="minorUnit"/>.
+        /// </summary>
+        /// <param name="currency">The currency to create a value.</param>
+        /// <param name="majorUnit">The major unit of the fungible asset value,
+        /// i.e., digits <em>before</em> the decimal separator.</param>
+        /// <param name="minorUnit">The minor unit of the fungible asset value,
+        /// i.e., digits <em>after</em> the decimal separator.</param>
+        /// <exception cref="ArgumentException">Thrown when the negativity sign is ambiguous
+        /// (e.g., both units have signs) or too big for the <paramref name="currency"/>.
+        /// </exception>
+        /// <seealso cref="Assets.Currency.DecimalPlaces"/>
+        public FungibleAssetValue(Currency currency, BigInteger majorUnit, BigInteger minorUnit)
+            : this(
+                currency,
+                majorUnit.IsZero ? minorUnit.Sign : majorUnit.Sign,
+                BigInteger.Abs(majorUnit),
+                BigInteger.Abs(minorUnit)
+            )
+        {
+            if (!majorUnit.IsZero && minorUnit < BigInteger.Zero)
+            {
+                throw new ArgumentException(
+                    "Unless the major unit is zero, the minor unit cannot be negative.",
+                    nameof(minorUnit)
+                );
+            }
+        }
+
+        /// <summary>
+        /// Creates a value of the <paramref name="currency"/> from the given <paramref
+        /// name="sign"/>, <paramref name="majorUnit"/> and <paramref name="minorUnit"/>.
+        /// </summary>
+        /// <param name="currency">The currency to create a value.</param>
+        /// <param name="sign">Indicates the sign (negative, positive, or zero) of the value.
```

Comment:
We may need new enum for this.

Reply:
Avoiding a new enum type was intended, because:

- [`BigInteger.Sign` property][1] does the same thing.
- It is much handier than enum values when adding a sign to values, i.e., you can simply do that by `v.Sign * anotherValue` instead of `v.Sign == Sign.Negative ? -anotherValue : anotherValue`.

[1]: https://docs.microsoft.com/en-us/dotnet/api/system.numerics.biginteger.sign?view=netstandard-2.0